src/platformdirs/windows.py

```diff
@@ -161,108 +170,172 @@ def get_win_folder_from_env_vars(csidl_name: str) -> str:
+        """
+        Get folder from the registry.
 
-    # Downgrade to short path name if it has high-bit chars.
-    if any(ord(c) > 255 for c in buf):  # noqa: PLR2004
-        buf2 = ctypes.create_unicode_buffer(1024)
-        if windll.kernel32.GetShortPathNameW(buf.value, buf2, 1024):
-            buf = buf2
+        This is a fallback technique at best. I'm not sure if using the registry for these guarantees us the correct
+        answer for all FOLDERID_* names.
 
-    if csidl_name == "CSIDL_DOWNLOADS":
-        return os.path.join(buf.value, "Downloads")  # noqa: PTH118
+        """
+        shell_folder_name = {
+            "FOLDERID_RoamingAppData": "AppData",
+            "FOLDERID_ProgramData": "Common AppData",
+            "FOLDERID_LocalAppData": "Local AppData",
+            "FOLDERID_Documents": "Personal",
+            "FOLDERID_Downloads": "{" + FOLDERID_Downloads_guid_string + "}",
+            "FOLDERID_Pictures": "My Pictures",
+            "FOLDERID_Videos": "My Video",
+            "FOLDERID_Music": "My Music",
+        }.get(folderid_name)
+        if shell_folder_name is None:
+            msg = f"Unknown FOLDERID name: {folderid_name}"
+            raise ValueError(msg)
+        if sys.platform != "win32":  # only needed for mypy type checker to know that this code runs only on Windows
+            raise NotImplementedError
+
+        key = winreg.OpenKey(
+            winreg.HKEY_CURRENT_USER, r"Software\Microsoft\Windows\CurrentVersion\Explorer\Shell Folders"
+        )
+        directory, _ = winreg.QueryValueEx(key, shell_folder_name)
+        return str(directory)
+
+
+if "ctypes" in globals() and hasattr(ctypes, "windll"):
+
+    class GUID(ctypes.Structure):
+        """
+        `
+        The GUID structure from Windows's guiddef.h header
+        <https://learn.microsoft.com/en-us/windows/win32/api/guiddef/ns-guiddef-guid>`_.
+        """
 
-    return buf.value
+        Data4Type = ctypes.c_ubyte * 8
+
+        _fields_ = (
+            ("Data1", ctypes.c_ulong),
+            ("Data2", ctypes.c_ushort),
+            ("Data3", ctypes.c_ushort),
+            ("Data4", Data4Type),
+        )
+
+        def __init__(self, guid_string: str) -> None:
+            digit_groups = guid_string.split("-")
+            expected_digit_groups = 5
+            if len(digit_groups) != expected_digit_groups:
+                msg = f"The guid_string, {guid_string!r}, does not contain {expected_digit_groups} groups of digits."
+                raise ValueError(msg)
+            for digit_group, expected_length in zip(digit_groups, (8, 4, 4, 4, 12)):
+                if len(digit_group) != expected_length:
+                    msg = (
+                        f"The digit group, {digit_group!r}, in the guid_string, {guid_string!r}, was the wrong length. "
+                        f"It should have been {expected_length} digits long."
+                    )
+                    raise ValueError(msg)
```

Comment:
Although I appreciate the thoroughness of the edge case handling, these checks are superfluous because we control the inputs. Let's remove these for better performance.